src/core/Form/Combobox/Combobox/Combobox.tsx

```diff
@@ -0,0 +1,595 @@
+import { ChipList } from '../ChipList/ChipList';
+import { baseStyles } from './Combobox.baseStyles';
+
+const baseClassName = 'fi-combobox';
+const comboboxClassNames = {
+  wrapper: `${baseClassName}_wrapper`,
+  open: `${baseClassName}--open`,
+  removeAllButton: `${baseClassName}_removeAllButton`,
+  error: `${baseClassName}--error`,
+};
+
+export interface ComboboxData {
+  /** Unique label that will be shown on combobox item and used on filter */
+  labelText: string;
+  /** Using labelText if chipText is not given */
+  chipText?: string;
+  /** Item selection disabled for the user */
+  disabled?: boolean;
+  /** Unique id to identify the item */
+  uniqueItemId: string;
+}
+
+export interface ComboboxProps<T extends ComboboxData> {
+  /** Combobox container div class name for custom styling. */
+  className?: string;
+  /** Items for the combobox */
+  items: Array<T & ComboboxData>;
+  /**
+   * Unique id
+   * If no id is specified, one will be generated automatically
+   */
+  id?: string;
+  /** Label */
+  labelText: string;
+  /** Event that is sent when item selections change */
+  onItemSelectionsChange?: (selectedItems: Array<T>) => void;
+  /** Show chip list */
+  chipListVisible?: boolean;
+  /** Chip action label */
+  chipActionLabel?: string;
+  /** Label for remove button. If it is given, button will be shown. */
+  removeAllButtonLabel?: string;
+  /** Placeholder text for input. Use only as visual aid, not for instructions. */
+  visualPlaceholder?: string;
+  /** Label to show when no items to show, e.g filtered all out */
+  emptyItemsLabel: string;
+  /** Default selected items */
+  defaultSelectedItems?: Array<T & ComboboxData>;
+  /** Event sent when filter changes */
+  onChange?: (value: string | undefined) => void;
+  /** Debounce time in milliseconds for onChange function. No debounce is applied if no value is given. */
+  debounce?: number;
+  /**
+   * 'default' | 'error'
+   * @default default
+   */
+  status?: FilterInputStatus;
+  /** Status text to be shown below the component and hint text. Use e.g. for validation error */
+  statusText?: string;
+  /** Controlled items; if item is in array, it is selected. If item has disabled: true, it will be disabled. */
+  selectedItems?: Array<T & ComboboxData>;
+  /** Selecting the item will send event with the id */
+  onItemSelect?: (uniqueItemId: string) => void;
+  /** Event to be sent when pressing remove all button */
+  onRemoveAll?: () => void;
+}
+
+// actual boolean value does not matter, only if it exists on the list
+interface ItemKeys {
+  [key: string]: boolean;
+}
+
+interface ComboboxState<T extends ComboboxData> {
+  filterInputValue: string | undefined;
+  filteredItems: T[];
+  showPopover: boolean;
+  currentSelection: string | null;
+  selectedKeys: ItemKeys;
+  selectedItems: T[];
+  initialItems: T[];
+  disabledKeys: ItemKeys;
+}
+
+function getSelectedKeys<T>(items: (T & ComboboxData)[] = []): ItemKeys {
```

Comment:
Would these be more efficient if there was just getItemKeys which would iterate all items and set selected and disabled values? Accessing items[key].disabled should be as efficient as accessing disabledItems[key] regardless of number of items in the objects.

Reply:
Not sure about the efficiency here. By quickly glancing that causes many changes to the current logic and therefore possibly increasing complexity in many places. So the efficiency gained in here vs the efficient lost in those other places seems unwanted.
I suggest that this can be improved in later PR if it is a real problem and we need to optimise it.